irr/src/SkinnedMesh.cpp

```diff
@@ -427,85 +340,102 @@ void SkinnedMesh::topoSortJoints()
 	for (u16 i = 0; i < n; ++i)
 		old_to_new_id[new_to_old_id[i]] = i;
 
-	std::vector<SJoint *> joints(n);
+	std::vector<SJoint *> sorted_joints(n);
 	for (u16 i = 0; i < n; ++i) {
-		joints[i] = AllJoints[new_to_old_id[i]];
-		joints[i]->JointID = i;
-		if (auto parentId = joints[i]->ParentJointID)
-			joints[i]->ParentJointID = old_to_new_id[*parentId];
+		auto *joint = joints[new_to_old_id[i]];
+		if (auto parentId = joint->ParentJointID)
+			joint->ParentJointID = old_to_new_id[*parentId];
+		sorted_joints[i] = joint;
+		joint->JointID = i;
 	}
-	AllJoints = std::move(joints);
-
+	// Verify that the topological ordering is correct
 	for (u16 i = 0; i < n; ++i) {
-		if (auto pjid = AllJoints[i]->ParentJointID)
+		if (auto pjid = sorted_joints[i]->ParentJointID)
 			assert(*pjid < i);
 	}
+	getJoints() = std::move(sorted_joints);
+
+	for (auto &weight : weights) {
+		weight.joint_id = old_to_new_id[weight.joint_id];
+	}
 }
 
 //! called by loader after populating with mesh and bone data
-SkinnedMesh *SkinnedMeshBuilder::finalize()
+SkinnedMesh *SkinnedMeshBuilder::finalize() &&
 {
 	os::Printer::log("Skinned Mesh - finalize", ELL_DEBUG);
 
 	topoSortJoints();
 
-	// Set array sizes
-	for (u32 i = 0; i < LocalBuffers.size(); ++i) {
-		Vertices_Moved.emplace_back(LocalBuffers[i]->getVertexCount());
-	}
-
-	prepareForSkinning();
+	mesh->prepareForSkinning();
 
 	std::vector<core::matrix4> matrices;
-	matrices.reserve(AllJoints.size());
-	for (auto *joint : AllJoints) {
+	matrices.reserve(getJoints().size());
+	for (auto *joint : getJoints()) {
 		if (const auto *matrix = std::get_if<core::matrix4>(&joint->transform))
 			matrices.push_back(*matrix);
 		else
 		 	matrices.push_back(std::get<core::Transform>(joint->transform).buildMatrix());
 	}
-	calculateGlobalMatrices(matrices);
+	mesh->calculateGlobalMatrices(matrices);
 
-	for (size_t i = 0; i < AllJoints.size(); ++i) {
-		auto *joint = AllJoints[i];
+	for (size_t i = 0; i < getJoints().size(); ++i) {
+		auto *joint = getJoints()[i];
 		if (!joint->GlobalInversedMatrix) {
 			joint->GlobalInversedMatrix = matrices[i];
 			joint->GlobalInversedMatrix->makeInverse();
 		}
 		// rigid animation for non animated meshes
 		for (u32 attachedMeshIdx : joint->AttachedMeshes) {
-			SSkinMeshBuffer *Buffer = (*SkinningBuffers)[attachedMeshIdx];
+			SSkinMeshBuffer *Buffer = (*mesh->SkinningBuffers)[attachedMeshIdx];
 			Buffer->Transformation = matrices[i];
 		}
 	}
 
-	recalculateBaseBoundingBoxes();
-	StaticPoseBox = calculateBoundingBox(matrices);
+	for (const auto &weight : weights) {
+		auto *buf = mesh->LocalBuffers.at(weight.buffer_id);
+		if (!buf->Weights)
+			buf->Weights = WeightBuffer(buf->getVertexCount());
+		if (weight.vertex_id >= buf->Weights->size()) {
+			throw std::out_of_range("vertex id out of range");
```

Comment:
This should be checked in `addWeight`, either as exception or as debug `assert`.